src/Proto3/Wire/Decode.hs

```diff
@@ -127,7 +127,26 @@ data ParsedField = VarintField Word64
 -- fromList [(1,[6,3]),(2,[4])]
 --
 toMap :: [(FieldNumber, v)] -> M.IntMap [v]
-toMap kvs0 = M.fromListWith (<>) . map (fmap (:[])) . map (first (fromIntegral . getFieldNumber)) $ kvs0
+toMap kvs0 = makeMap . map (first (fromIntegral . getFieldNumber)) $ kvs0
+  where
+    makeMap :: [(Int, v)] -> M.IntMap [v]
+    makeMap = close . foldl' combineSeen (M.empty, Nothing)
+
+    close (m, Nothing) = m
+    close (m, Just (k, v)) = M.insert k v m
+
+    -- If keys are in order, then we don't have to make any lookups,
+    -- we just maintain the active element.
+    -- Out of order keys will lookup in the map
+    combineSeen :: (M.IntMap [v], Maybe (Int, [v])) -> (Int, v) -> (M.IntMap [v], Maybe (Int, [v]))
+    combineSeen (_, Nothing) (k1, a1) = (M.empty, Just (k1, [a1]))
```

Comment:
The main thing I'd suggest is to change the accumulator type from `(IntMap [v], Maybe (Int, [v]))` to `Maybe (IntMap [v], Int, [v])` since as far as I can tell the `IntMap` is supposed not be used if the `Maybe` is `Nothing`, so you want to make invalid states unrepresentable

Reply:
I implemented this change in 91abc60